Flaky test TestBinPackingNodeAllocator.testNoMatchingNodeAvailableTimeoutReset

https://github.com/trinodb/trino/actions/runs/4675961675/jobs/8282707441 (https://github.com/trinodb/trino/pull/16970)

```
2023-04-12T03:43:56.067-0600	ERROR	pool-24-thread-2	io.trino.testng.services.ProgressLoggingListener	[TEST FAILURE] io.trino.execution.scheduler.TestBinPackingNodeAllocator.testNoMatchingNodeAvailableTimeoutReset; (took: 0.0 seconds)
java.lang.AssertionError: node lease acquired expected [false] but found [true]
	at org.testng.Assert.fail(Assert.java:94)
	at org.testng.Assert.failNotEquals(Assert.java:513)
	at org.testng.Assert.assertFalse(Assert.java:63)
	at io.trino.execution.scheduler.TestBinPackingNodeAllocator.assertNotAcquired(TestBinPackingNodeAllocator.java:678)
	at io.trino.execution.scheduler.TestBinPackingNodeAllocator.testNoMatchingNodeAvailableTimeoutReset(TestBinPackingNodeAllocator.java:385)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)

[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:    TestBinPackingNodeAllocator.testNoMatchingNodeAvailableTimeoutReset:385->assertNotAcquired:678 node lease acquired expected [false] but found [true]
[INFO] 
Error:  Tests run: 7996, Failures: 1, Errors: 0, Skipped: 0
[INFO] 
```

--- a/core/trino-main/src/main/java/io/trino/metadata/InMemoryNodeManager.java
+++ b/core/trino-main/src/main/java/io/trino/metadata/InMemoryNodeManager.java
@@ -17,6 +17,8 @@ import com.google.common.collect.ImmutableSet;
 import io.trino.client.NodeVersion;
 import io.trino.spi.connector.CatalogHandle;
 
+import javax.annotation.concurrent.ThreadSafe;
+
 import java.net.URI;
 import java.util.Optional;
 import java.util.Set;
@@ -25,6 +27,7 @@ import java.util.function.Consumer;
 
 import static java.util.Objects.requireNonNull;
 
+@ThreadSafe
 public class InMemoryNodeManager
         implements InternalNodeManager
 {

--- a/core/trino-main/src/main/java/io/trino/metadata/InMemoryNodeManager.java
+++ b/core/trino-main/src/main/java/io/trino/metadata/InMemoryNodeManager.java
@@ -62,7 +62,7 @@ public class InMemoryNodeManager
     {
         switch (state) {
             case ACTIVE:
-                return allNodes;
+                return ImmutableSet.copyOf(allNodes);
             case INACTIVE:
             case SHUTTING_DOWN:
                 return ImmutableSet.of();
@@ -73,20 +73,20 @@ public class InMemoryNodeManager
     @Override
     public Set<InternalNode> getActiveCatalogNodes(CatalogHandle catalogHandle)
     {
-        return allNodes;
+        return ImmutableSet.copyOf(allNodes);
     }
 
     @Override
     public NodesSnapshot getActiveNodesSnapshot()
     {
-        return new NodesSnapshot(allNodes, Optional.empty());
+        return new NodesSnapshot(ImmutableSet.copyOf(allNodes), Optional.empty());
     }
 
     @Override
     public AllNodes getAllNodes()
     {
         return new AllNodes(
-                allNodes,
+                ImmutableSet.copyOf(allNodes),
                 ImmutableSet.of(),
                 ImmutableSet.of(),
                 ImmutableSet.of(CURRENT_NODE));

--- a/core/trino-main/src/main/java/io/trino/metadata/InternalNodeManager.java
+++ b/core/trino-main/src/main/java/io/trino/metadata/InternalNodeManager.java
@@ -22,6 +22,7 @@ import java.util.Optional;
 import java.util.Set;
 import java.util.function.Consumer;
 
+import static com.google.common.base.MoreObjects.toStringHelper;
 import static java.util.Objects.requireNonNull;
 
 public interface InternalNodeManager
@@ -67,6 +68,15 @@ public interface InternalNodeManager
             return connectorNodes
                     .map(map -> map.get(catalogHandle))
                     .orElse(allNodes);
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("allNodes", allNodes)
+                    .add("connectorNodes", connectorNodes.orElse(null))
+                    .toString();
         }
     }
 }

--- a/core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java
+++ b/core/trino-main/src/test/java/io/trino/execution/scheduler/TestBinPackingNodeAllocator.java
@@ -40,6 +40,7 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.TimeUnit;
 
 import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;
 import static io.airlift.concurrent.MoreFutures.getFutureValue;
 import static io.airlift.units.DataSize.Unit.GIGABYTE;
 import static io.trino.testing.TestingHandles.createTestCatalogHandle;
@@ -380,6 +381,10 @@ public class TestBinPackingNodeAllocator
 
             // remove the node - we are again in situation where no matching nodes exist in cluster
             nodeManager.removeNode(NODE_2);
+
+            // sleep for a while before releasing lease, as background processPendingAcquires may be still running with old snapshot
+            // containing NODE_2, and theNotAcquireLease could be fulfilled when theAcquireLease is released
+            sleepUninterruptibly(10, TimeUnit.MILLISECONDS);
             theAcquireLease.release();
             nodeAllocatorService.processPendingAcquires();
             assertNotAcquired(theNotAcquireLease);

